File: Leetcode_206_Reverse_Linked_List.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author KD
 */
public class Leetcode_206_Reverse_Linked_List {
    public ListNode reverseList(ListNode head){
        ListNode node = head;
        ListNode prev = null;
        while(node != null){
            ListNode nextNode = node.next;
            node.next = prev;
            prev = node;
            node = nextNode;
        }
        return prev;
    }
}

File: Leetcode_206. Reverse Linked List.py
'''
206. Reverse Linked List

Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?
'''


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# class Solution:
#     def reverseList(self, head):
#         """
#         :type head: ListNode
#         :rtype: ListNode
#         """
#         node = head
#         save = None
#         while node:
#             nextNode = node.next
#             node.next = save
#             save = node
#             node = nextNode
#         return save


class Solution:
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next: return head
        nextNode = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return nextNode
